saviSimulation/savi/commandStation/SocketConnector.java

```diff
@@ -0,0 +1,100 @@
+package savi.commandStation;
+
+import java.io.IOException;
+import java.net.DatagramPacket;
+import java.net.DatagramSocket;
+import java.net.InetAddress;
+import java.net.SocketException;
+import java.net.UnknownHostException;
+
+public class SocketConnector {
+	
+	private int myPort;
+	private int otherPort;
+	
+	private MessageHandler inboxHandler;
+	
+
+	
+	public SocketConnector(MessageHandler msgHandler, int myPort,
+			int otherPort) {
+		
+		inboxHandler = msgHandler;
+		this.myPort = myPort;
+		this.otherPort = otherPort;
+		
+	}
+
+
+
+	public void messageOut(String msg) {
+		System.out.println("Message OUT: "+msg);
+		//get data for message
+		byte [] data = msg.getBytes();
+         
+        try {
+        	DatagramSocket socket = new DatagramSocket();
+        	socket.setBroadcast(true);
+
+        	InetAddress aHost = InetAddress.getLocalHost();
+
+        	DatagramPacket packet = new DatagramPacket(
+        			data,
+        			data.length,
+        			aHost,
+        			otherPort);
+
+        	socket.send(packet);
+        	socket.close();
+
+        } catch (SocketException e) {
+            e.printStackTrace();
+        } catch (UnknownHostException e) {			
+			e.printStackTrace();
+		} catch (IOException e) {
+			e.printStackTrace();
+		} 
+
+	}
+	
+	/***
+	 * listens for incoming messages on port commandStationPort
+	 */
+	public void listenForMessages() {
+	    (new Thread() {
+	        @Override
+	        public void run() {
+
+	                //byte data[] = new byte[0];
+	                DatagramSocket socket = null;
+	                try {
+	                    socket = new DatagramSocket(myPort);
```

Comment:
I see a warning on this line in my IDE - this socket is never closed leading to a possible resource leak. I suspect that this is not an issue for us as we open the socket when the program starts and let the garbage collection close them. Am I correctly understanding this?

Reply:
I'm not exactly sure, but I do believe the garbage collector should close and free any necessary resources. With this being said, we shouldn't rely on the garbage collector to do this job for us (it doesn't always have predictable behaviour). It is good practice to explicitly close the socket when we no longer need it

Reply:
So as it is the socket is used inside an infinite loop and we only leave it when the program ends. If we wanted to shut down the program more cleanly then we would need to add some code to do this, including interrupting the socket.receive() method which is waiting for a packet (or send a shutdown packet). Certainly it's feasible but unless there's clear perspectives of this work moving along with Cohort I wouldn't invest the time in that sort of enhancements.

Reply:
I concur with this view, I just wanted to make sure that we were all on the same page here.

All of that said, we could try to set things up so that the control station could connect and disconnect to the simulation at any time during execution without preventing the simulation from continuing... although I don't think this is worth it at this stage. My view is that the socket piece is a bonus for now.

Reply:
It currently works as you describe. You can run the simulation without the control station (except that the agents don't do anything since they're waiting for orders). Starting / stopping the command station doesn't affect the simulation.